### PR TITLE
[Text Generation][KVCacheStorage] `ChatPipeline` implementation

### PR DESCRIPTION
**Pull request by dbogunowicz**
Implementation of `ChatPipeline` that builds on top of `TextGenerationPipeline`, but uses session_ids and StorageKVCache to enable continuous recollection of past kv_cache information.

Example use:
```python
from deepsparse import Pipeline

pipeline = Pipeline.create(
    task="chat",
    model_path="/home/ubuntu/damian/sparseml/deployment_opt",
)
session_id = "session_id"
while True:
    # get input from user
    input_text = input("User: ")
    response = pipeline(sequences=[input_text], session_ids=session_id, max_tokens=32)
    print("Bot: ", response.generations[0].text)
```

```bash
User: Hi my name is Damian and I am from Italy.
Bot:   I am a professional photographer and I have been working in the industry for over 10 years. I have been working in the industry for the last 5 years and I
User: am from   
Bot:   Italy. I have been working in the industry for the last 5 years and I am from  Italy. I have been working in the industry for the last 5
User: Did I mention that my name is
Bot:   Damian?

I am a professional photographer and I have been working in the industry for the last 5 years and 
```

Testing:
- successfully ran the full LLM testing suite

### CODE REVIEW

File: src/deepsparse/transformers/pipelines/chat.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        If the storage kv cache does not have a kv cache decoder,
+        create a new one and store it in the storage kv cache.
+
+        :param engine_inputs: the inputs for the engine
+        :return: the initialized kv cache decoder
+        """
+        session_id = engine_inputs[-1]
+        kv_cache = self.storage_kv_cache.get(session_id)
+        if kv_cache is None:
+            kv_cache = super().get_kv_cache_decoder(engine_inputs)
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/chat.py:152]: nice! let's add an in-line comment w/ the expected behavior of the super fn

File: src/deepsparse/transformers/pipelines/chat.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+        :param engine_input: the inputs for the engine
+        :param inputs: the input schema for the pipeline
+
+        :return: the engine input with the session ids
+        """
+        session_ids = inputs.session_ids
+        if session_ids is None:
+            # session_ids is None, so we need to generate
+            # a session id for each input sequence
+            # TODO: Talk to Dipika whether this aligns with the
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/chat.py:173]: I don't think this will - we'll need to do a bit more planning here - cc @dsikka

File: src/deepsparse/transformers/pipelines/chat.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    )
+
+
+@Pipeline.register(
+    task="chat",
+    task_aliases=["chatbot"],
+)
+class ChatPipeline(TextGenerationPipeline):
+    """
+    Pipeline for chat applications using transformers models.
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/chat.py:71]: add lifecycle list that shows the updates this makes to the base TG pipeline

File: src/deepsparse/transformers/utils/storage_kv_cache.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+import logging
+
+
+_LOGGER = logging.getLogger(__name__)
+
+__all__ = ["SessionStorageKVCache"]
+
+
+class SessionStorageKVCache(dict):
```
Comment (bfineran) [inline src/deepsparse/transformers/utils/storage_kv_cache.py:23]: do we want to override any default methods to include more validation? additionally, we may want to add a capacity/memory to the dict eventually (ie LRU caching) might want to think a bit more on if this is the way to go

**Review by bfineran — COMMENTED** [review]
let's also make sure to add the tests back in. happy with how this is looking

File: src/deepsparse/transformers/utils/storage_kv_cache.py (hunk 4 above)
Comment (dbogunowicz) [reply src/deepsparse/transformers/utils/storage_kv_cache.py:23]: The new design simplifies so many things that I would not do any (over)engineering here. So far, all we need from this object is the ability to assign key-value pairs and fetch a value given a key. In the future, we might expand this class, but for now, I do not see a good reason.

File: src/deepsparse/transformers/utils/helpers.py [hunk 5]
```diff
@@ -28,11 +28,50 @@
     "repeat_inputs",
     "initialize_kv_cache_state",
     "prepends_bos_token",
+    "validate_session_ids",
 ]
 
 _LOGGER = logging.getLogger(__name__)
 
 
+def validate_session_ids(
```
Comment (dbogunowicz) [inline src/deepsparse/transformers/utils/helpers.py:37]: @bfineran still want to sync on the edge cases here?

File: src/deepsparse/utils/data.py [hunk 6]
```diff
@@ -168,7 +168,7 @@ def numpy_softmax(x: numpy.ndarray, axis: int = 0):
 
 def split_engine_inputs(
```
Comment (dbogunowicz) [inline src/deepsparse/utils/data.py:169]: just a cleanup of this function, not really part of this PR

File: tests/deepsparse/transformers/pipelines/test_text_generation.py [hunk 7]
```diff
@@ -22,7 +22,7 @@
 from tests.deepsparse.transformers.pipelines.helpers import TorchGroundTruthSource
 
 
-_PRECISION = 1e-3
+_PRECISION = 1e-4
```
Comment (dbogunowicz) [inline tests/deepsparse/transformers/pipelines/test_text_generation.py:25]: required by the OPT-1.3b

File: tests/deepsparse/transformers/pipelines/test_text_generation.py [hunk 8]
```diff
@@ -51,6 +51,10 @@ def Fibonacci(n):
     "internal_kv_cache",
     [True, False],
 )
+@pytest.mark.parametrize(
```
Comment (dbogunowicz) [inline tests/deepsparse/transformers/pipelines/test_text_generation.py:54]: this is kind of ugly but makes do for now. we scan still quickly test the new chat scenario. refactor for the tests pending.

File: tests/deepsparse/transformers/pipelines/test_text_generation.py (hunk 7 above)
Comment (dbogunowicz) [inline tests/deepsparse/transformers/pipelines/test_text_generation.py:25]:
```suggestion
_PRECISION = 1e-3
```

File: src/deepsparse/transformers/pipelines/chat.py [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+
+    def add_session_ids_to_engine_input(
+        self, engine_input: List[numpy.ndarray], inputs: ChatInput
+    ):
+        """
+        Create the session ids for the input sequences (if not provided)
+        and add them to the engine input.
+
+        :param engine_input: the inputs for the engine
+        :param inputs: the input schema for the pipeline
```
Comment (rahul-tuli) [inline src/deepsparse/transformers/pipelines/chat.py:231]: nit: extra new line

File: src/deepsparse/transformers/pipelines/chat.py [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+        return batch_outputs
+
+    def split_engine_inputs(
+        self, items: List[Union[numpy.ndarray, List[str]]], batch_size: int
+    ) -> Tuple[List[List[numpy.ndarray]], int]:
+        """
+        Custom implementation of splitting the engine inputs that takes into
+        account the fact that the `items` contain additionally a list of
+        session_ids, that need to be distributed across the batches.
+        :param items: list of numpy arrays to split (plus list of session_ids)
```
Comment (rahul-tuli) [inline src/deepsparse/transformers/pipelines/chat.py:278]: new line

File: src/deepsparse/transformers/pipelines/chat.py [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+        :param batch_size: size of each batch to split into
+        :return: list of batches, where each batch is a list of numpy arrays
+            (plus session_ids), as well as the total batch size
+        """
+        # extract the session_ids from the items
+        session_ids = next((item for item in items if isinstance(item, list)), None)
```
Comment (rahul-tuli) [inline src/deepsparse/transformers/pipelines/chat.py:284]: nice

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 12]
```diff
@@ -572,9 +574,11 @@ def engine_forward(
 
         :param engine_inputs: list of numpy inputs to
             Pipeline engine forward pass
-        :return: A tuple of numpy array that contains the
-            sequence of generated tokens and a sequence
-            of logits for each generated token
+        :return: A tuple of:
+            - numpy array that contains the sequence
+                of generated tokens
+            - numpy array that cointains the sequence of
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/text_generation.py:580]: is this a tangential change?

Comment (bfineran) [reply src/deepsparse/transformers/pipelines/text_generation.py:580]: also looks out of date after rebase

Comment (dbogunowicz) [reply src/deepsparse/transformers/pipelines/text_generation.py:580]: Oh yes, this is a tangential change, sorry, you are right. Yes, it is out of date.
